caffe2/sgd/adadelta_op_gpu.cu

```diff
@@ -0,0 +1,168 @@
+#include "adadelta_op.h"
+#include "caffe2/core/common_gpu.h"
+#include "caffe2/core/context_gpu.h"
+#include "caffe2/utils/mixed_utils.h"
+
+namespace caffe2 {
+
+__global__ void AdadeltaUpdate(
+    int N,
+    const float* w,
+    const float* g,
+    const float* h,
+    const float* d,
+    float* nw,
+    float* nh,
+    float* nd,
+    float epsilon,
+    float decay,
+    const float* lr) {
+  CUDA_1D_KERNEL_LOOP(i, N) {
+    float gi = g[i];
+    float di = d[i];
+    float hi = nh[i] = decay * h[i] + (1.0f - decay) * gi * gi;
+    float ng = (std::sqrt(di + epsilon) / std::sqrt(hi + epsilon)) * gi;
+    nw[i] = w[i] + lr[0] * ng;
+    nd[i] = decay * di + (1.0f - decay) * ng * ng;
+  }
+}
+
+template <>
+void adadelta_update<CUDAContext>(
+    int N,
+    const float* w,
+    const float* g,
+    const float* h,
+    const float* d,
+    float* nw,
+    float* nh,
+    float* nd,
+    float epsilon,
+    float decay,
+    const float* lr,
+    CUDAContext* context) {
+  AdadeltaUpdate<<<
+      CAFFE_GET_BLOCKS(N),
+      CAFFE_CUDA_NUM_THREADS,
+      0,
+      context->cuda_stream()>>>(N, w, g, h, d, nw, nh, nd, epsilon, decay, lr);
+}
+
+template <typename SIndex, typename THalf>
+__global__ void SparseAdadeltaKernel(
+    const size_t N,
+    const size_t grad_slice_sz,
+    const float epsilon,
+    const float decay,
+    THalf* param,
+    THalf* param_mom,
+    THalf* param_mom_delta,
+    const SIndex* indices,
+    const float* grad,
+    const float* lr) {
+  const float LR = lr[0];
+  CUDA_1D_KERNEL_LOOP(i, N) {
+    const size_t gradIdx = i;
+    const SIndex index = indices[i / grad_slice_sz];
+    const size_t paramIdx = index * grad_slice_sz + (i % grad_slice_sz);
+
+    float mom_new = mixed_mult(decay, param_mom[paramIdx]) +
+        (1.0f - decay) * grad[gradIdx] * grad[gradIdx];
+    mixed_store(&mom_new, &(param_mom[paramIdx]));
+    float grad_new = (sqrt(mixed_add(epsilon, param_mom_delta[paramIdx])) /
+                      sqrt(mom_new + epsilon)) *
+        grad[gradIdx];
+    float param_new = mixed_add(LR * grad_new, param[paramIdx]);
+    mixed_store(&param_new, &(param[paramIdx]));
+    float mom_delta_new = mixed_mult(decay, param_mom_delta[paramIdx]) +
+        (1.0f - decay) * grad_new * grad_new;
+    mixed_store(&mom_delta_new, &(param_mom_delta[paramIdx]));
+  }
+}
+
+template <typename T, class Context>
+class CUDASparseAdadeltaOp final : public Operator<Context> {
+ public:
+  USE_OPERATOR_CONTEXT_FUNCTIONS;
+  CUDASparseAdadeltaOp(const OperatorDef& operator_def, Workspace* ws)
+      : Operator<Context>(operator_def, ws),
+        epsilon_(OperatorBase::GetSingleArgument<T>("epsilon", 1e-5f)),
```

Comment:
Here we can make epsilon and decay as float type and cast it to T when using it.